File: src/main/java/com/badbones69/crazyenchantments/CrazyEnchantments.java
package com.badbones69.crazyenchantments;

import com.badbones69.crazyenchantments.api.FileManager.Files;
import com.badbones69.crazyenchantments.api.PluginSupport.SupportedPlugins;
import com.badbones69.crazyenchantments.api.support.misc.spawners.SilkSpawnerSupport;
import com.badbones69.crazyenchantments.commands.*;
import com.badbones69.crazyenchantments.controllers.*;
import com.badbones69.crazyenchantments.enchantments.*;
import com.badbones69.crazyenchantments.listeners.*;
import org.bstats.bukkit.Metrics;
import org.bukkit.command.CommandExecutor;
import org.bukkit.command.PluginCommand;
import org.bukkit.command.TabCompleter;
import org.bukkit.configuration.file.FileConfiguration;
import org.bukkit.event.Listener;
import org.bukkit.plugin.PluginManager;
import org.bukkit.plugin.java.JavaPlugin;

public class CrazyEnchantments extends JavaPlugin implements Listener {

    private static CrazyEnchantments plugin;

    private Starter starter;

    // Plugin Listeners.
    public PluginManager pluginManager = getServer().getPluginManager();

    private FireworkDamageListener fireworkDamageListener;
    private ShopListener shopListener;

    private ArmorEnchantments armorEnchantments;

    // Menus.
    private Tinkerer tinkerer;
    private BlackSmith blackSmith;
    private GKitzController gKitzController;

    @Override
    public void onEnable() {
        plugin = this;

        starter = new Starter();

        // Create all instances we need.
        starter.run();

        starter.getCurrencyAPI().loadCurrency();

        FileConfiguration config = Files.CONFIG.getFile();

        String metricsPath = config.getString("Settings.Toggle-Metrics");
        boolean metricsEnabled = config.getBoolean("Settings.Toggle-Metrics");

        if (metricsPath == null) {
            config.set("Settings.Toggle-Metrics", false);

            Files.CONFIG.saveFile();
        }

        if (metricsEnabled) new Metrics(this, 4494);

        pluginManager.registerEvents(blackSmith = new BlackSmith(), this);
        pluginManager.registerEvents(tinkerer = new Tinkerer(), this);

        if (starter.getCrazyManager().isGkitzEnabled()) {
            getLogger().info("G-Kitz support is now enabled.");

            pluginManager.registerEvents(gKitzController = new GKitzController(), this);
        }

        pluginManager.registerEvents(fireworkDamageListener = new FireworkDamageListener(), this);
        pluginManager.registerEvents(shopListener = new ShopListener(), this);

        enable();
    }

    @Override
    public void onDisable() {
        disable();
    }

    private void enable() {
        // Load what we need to properly enable the plugin.
        starter.getCrazyManager().load();

        pluginManager.registerEvents(new DustControlListener(), this);

        pluginManager.registerEvents(new PickaxeEnchantments(), this);
        pluginManager.registerEvents(new HelmetEnchantments(), this);
        pluginManager.registerEvents(new SwordEnchantments(), this);
        pluginManager.registerEvents(armorEnchantments = new ArmorEnchantments(), this);
        pluginManager.registerEvents(new AllyEnchantments(), this);
        pluginManager.registerEvents(new ToolEnchantments(), this);
        pluginManager.registerEvents(new BootEnchantments(), this);
        pluginManager.registerEvents(new AxeEnchantments(), this);
        pluginManager.registerEvents(new BowEnchantments(), this);
        pluginManager.registerEvents(new HoeEnchantments(), this);

        pluginManager.registerEvents(new ProtectionCrystalListener(), this);
        pluginManager.registerEvents(new FireworkDamageListener(), this);
        pluginManager.registerEvents(new AuraListener(), this);

        pluginManager.registerEvents(new InfoGUIControl(), this);
        pluginManager.registerEvents(new LostBookController(), this);
        pluginManager.registerEvents(new CommandChecker(), this);

        if (SupportedPlugins.SILK_SPAWNERS.isCachedPluginLoaded()) {
            getLogger().fine("Silk Spawners support is enabled.");
            pluginManager.registerEvents(new SilkSpawnerSupport(), this);
        }

        if (SupportedPlugins.SILK_SPAWNERS_V2.isCachedPluginLoaded()) getLogger().fine("Silk Spawners support by Candc is enabled.");

        registerCommand(getCommand("crazyenchantments"), new CETab(), new CECommand());

        registerCommand(getCommand("tinkerer"), null, new TinkerCommand());
        registerCommand(getCommand("blacksmith"), null, new BlackSmithCommand());

        registerCommand(getCommand("gkit"), new GkitzTab(), new GkitzCommand());
    }

    private void disable() {
        armorEnchantments.stop();

        if (starter.getAllyManager() != null) starter.getAllyManager().forceRemoveAllies();

        getServer().getOnlinePlayers().forEach(starter.getCrazyManager()::unloadCEPlayer);
    }

    private void registerCommand(PluginCommand pluginCommand, TabCompleter tabCompleter, CommandExecutor commandExecutor) {
        if (pluginCommand != null) {
            pluginCommand.setExecutor(commandExecutor);

            if (tabCompleter != null) pluginCommand.setTabCompleter(tabCompleter);
        }
    }

    public static CrazyEnchantments getPlugin() {
        return plugin;
    }

    public Starter getStarter() {
        return starter;
    }

    // Plugin Listeners.
    public FireworkDamageListener getFireworkDamageListener() {
        return fireworkDamageListener;
    }

    public ShopListener getShopListener() {
        return shopListener;
    }

    public PluginManager getPluginManager() {
        return pluginManager;
    }

    public Tinkerer getTinkerer() {
        return tinkerer;
    }

    public BlackSmith getBlackSmith() {
        return blackSmith;
    }

    public GKitzController getgKitzController() {
        return gKitzController;
    }
}

File: src/main/java/com/badbones69/crazyenchantments/CrazyEnchantments.kt
package com.badbones69.crazyenchantments

import com.badbones69.crazyenchantments.api.CrazyManager
import com.badbones69.crazyenchantments.api.FileManager
import com.badbones69.crazyenchantments.api.FileManager.Files
import com.badbones69.crazyenchantments.api.PluginSupport.SupportedPlugins
import com.badbones69.crazyenchantments.api.economy.CurrencyAPI
import com.badbones69.crazyenchantments.api.multisupport.misc.spawners.SilkSpawnerSupport
import com.badbones69.crazyenchantments.commands.*
import com.badbones69.crazyenchantments.controllers.*
import com.badbones69.crazyenchantments.enchantments.*
import io.papermc.lib.PaperLib
import org.bstats.bukkit.Metrics
import org.bukkit.attribute.Attribute
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.player.PlayerJoinEvent
import org.bukkit.event.player.PlayerQuitEvent
import org.bukkit.plugin.java.JavaPlugin

class CrazyEnchantments : JavaPlugin(), Listener {

    private val crazyManager = CrazyManager.getInstance()

    private val fileManager = FileManager.getInstance()

    private var pluginEnabled = false

    private var armor: Armor? = null

    // Avoid using @this
    private val plugin = this

    private val generic = Attribute.GENERIC_MAX_HEALTH

    override fun onEnable() {

        if (!PaperLib.isPaper()) {
            logger.warning("====================================================")
            logger.warning(" " + this.name + " works better if you use Paper ")
            logger.warning(" as your server software.")
            logger.warning(" ")
            logger.warning(" Paper offers significant performance improvements,")
            logger.warning(" bug fixes, security enhancements and optional")
            logger.warning(" features for server owners to enhance their server.")
            logger.warning(" ")
            logger.warning(" All of your plugins should still work, and the")
            logger.warning(" Paper community will gladly help you fix any issues.")
            logger.warning("")
            logger.warning(" Join the Purpur Community @ https://purpurmc.org/discord")
            logger.warning("====================================================")
            logger.warning("A few features might not work on Spigot so be warned.")

            // getServer().getPluginManager().disablePlugin(this);
        }

        runCatching {
            crazyManager.loadPlugin(plugin)

            fileManager.logInfo(true).setup(this)

            val metricsEnabled = Files.CONFIG.file.getBoolean("Settings.Toggle-Metrics")

            if (Files.CONFIG.file.getString("Settings.Toggle-Metrics") != null) {
                if (metricsEnabled) Metrics(plugin, 4494)
            } else {
                logger.warning("Metrics was automatically enabled.")
                logger.warning("Please add Toggle-Metrics: false to the top of your config.yml")
                logger.warning("https://github.com/Crazy-Crew/Crazy-Crates/blob/main/src/main/resources/config.yml")
                logger.warning("An example if confused is linked above.")

                Metrics(plugin, 4494)
            }

            crazyManager.load()

        }.onFailure {

            logger.severe(it.message)

            it.stackTrace.forEach { e -> logger.severe(e.toString()) }

            pluginEnabled = false

            return
        }

        SupportedPlugins.printHooks()

        CurrencyAPI.loadCurrency()

        val patchHealth = Files.CONFIG.file.getBoolean("Settings.Reset-Players-Max-Health")

        plugin.server.onlinePlayers.forEach {
            crazyManager.loadCEPlayer(it)

            if (patchHealth) it.getAttribute(generic)?.baseValue = it.getAttribute(generic)?.baseValue!!
        }

        plugin.server.scheduler.runTaskTimerAsynchronously(plugin, Runnable {
            crazyManager.cePlayers.forEach { crazyManager.backupCEPlayer(it) }
        }, 5 * 20 * 60, 5 * 20 * 60)

        enable()

        pluginEnabled = true;
    }

    override fun onDisable() {
        if (!pluginEnabled) return

        disable()
    }

    private fun enable() {

        listOf(
            plugin,
            ShopControl(),
            InfoGUIControl(),
            LostBookController(),
            EnchantmentControl(),
            SignControl(),
            DustControl(),
            Tinkerer(),
            AuraListener(),
            ScrollControl(),
            BlackSmith(),
            ArmorListener(),
            ProtectionCrystal(),
            Scrambler(),
            CommandChecker(),
            FireworkDamage(),
            Bows(),
            Axes(),
            Tools(),
            Hoes(),
            Helmets(),
            PickAxes(),
            Boots(),
            Swords(),
            Armor().also { armor = it },
            AllyEnchantments()
        ).onEach {loop ->
            server.pluginManager.registerEvents(loop, plugin)
        }

        if (crazyManager.isGkitzEnabled) {
            logger.info("Gkitz support is now enabled.")
            server.pluginManager.registerEvents(GKitzController(), plugin)
        }

        if (SupportedPlugins.SILKSPAWNERS.isPluginLoaded()) {
            logger.info("Silk Spawners support is now enabled.")
            server.pluginManager.registerEvents(SilkSpawnerSupport(), plugin)
        }

        getCommand("crazyenchantments")?.setExecutor(CECommand())
        getCommand("crazyenchantments")?.tabCompleter = CETab()
        getCommand("tinkerer")?.setExecutor(TinkerCommand())
        getCommand("blacksmith")?.setExecutor(BlackSmithCommand())
        getCommand("gkit")?.setExecutor(GkitzCommand())
        getCommand("gkit")?.tabCompleter = GkitzTab()
    }

    private fun disable() {
        armor?.stop()

        if (crazyManager.allyManager != null) crazyManager.allyManager.forceRemoveAllies()

        server.onlinePlayers.forEach { crazyManager.unloadCEPlayer(it) }
    }

    @EventHandler
    fun onPlayerJoin(e: PlayerJoinEvent) {
        val player = e.player
        crazyManager.loadCEPlayer(player)
        crazyManager.updatePlayerEffects(player)

        val patchHealth = Files.CONFIG.file.getBoolean("Settings.Reset-Players-Max-Health")

        if (patchHealth) player.getAttribute(generic)?.baseValue = player.getAttribute(generic)?.baseValue!!
    }

    @EventHandler
    fun onPlayerQuit(e: PlayerQuitEvent) {
        crazyManager.unloadCEPlayer(e.player)
    }
}